File: Homework_2/Problem_1/problem1.py
NEGATIVE_INFINITY = -1e10

def maxCrossSum(array, start, middle, end):
    
    tmpSum = 0

    leftCrossingSum = NEGATIVE_INFINITY
    rightCrossingSum = NEGATIVE_INFINITY

    for i in range(middle, -1, -1):
        tmpSum += array[i]
        leftCrossingSum = max(leftCrossingSum, tmpSum)

    tmpSum = 0

    for i in range(middle, end+1):
        tmpSum += array[i]
        rightCrossingSum = max(rightCrossingSum, tmpSum)

    return max(leftCrossingSum + rightCrossingSum - array[middle], leftCrossingSum, rightCrossingSum)


def maxContigousSum(array, start, end):
    if(start == end):
        return array[start]
    
    middle = (start + end) // 2

    leftSum = maxContigousSum(array, start, middle)
    rightSum = maxContigousSum(array, middle+1, end)
    crossSum = maxCrossSum(array, start, middle, end)

    return max(leftSum, rightSum, crossSum)

def readTestFile(testPath):
    testFile = open(testPath)
    testList = []

    tmpList = []

    for line in testFile:
        if(line == "END"):
            break
        tmpList = []
        for element in line.split(' '):
            tmpList.append(int(element))
        testList.append(tmpList)

    return testList

if __name__ == "__main__":
    testList = readTestFile("./test.txt")

    for test in testList:
        maximumSum = maxContigousSum(test, 0, len(test)-1)
        print(maximumSum)

File: Homework_2/Problem_1/problem1.cpp
#include <iostream>
#define MIN -9999999

int max(int a, int b) {
    if (a > b) return a;
    else return b;
};

int max(int a, int b, int c) {
    return max(max(a, b), c);
};


int maxMidSum(int* a, int start, int mid, int end) {

    int sum = 0;
    int left_sum = MIN;

    for (int i = mid; i >= start; i--) {
        sum += a[i];
        if (sum > left_sum)
            left_sum = sum;
    }

    sum = 0;
    int right_sum = MIN;

    for (int i = mid ; i <= end; i++) {
        sum += a[i];
        if (sum > right_sum)
            right_sum = sum;
    }
    return max(left_sum + right_sum - a[mid], left_sum, right_sum);
};


int maxSum(int* a, int start, int end) {

    if (start == end)
        return a[start];

    int mid = (start + end) / 2;

    return max(maxSum(a, start, mid), maxSum(a, mid + 1, end), maxMidSum(a, start, mid, end));
};



int main()
{
    
    int array[] = { 13, -3, -25, 20, -3, -16, -23, 18, 20, -7, 12, -5, -22, 15, -4, 7 };

    int n = sizeof(array) / sizeof(int);

    std::cout << "max sum is: " << maxSum(array, 0, n - 1);

}
